method/scanpy/method_scanpy.py

```diff
@@ -0,0 +1,185 @@
+observation_file = args.observations
+
+if args.neighbors is not None:
+    neighbors_file = args.neighbors
+if args.matrix is not None:
+    matrix_file = args.matrix
+if args.dim_red is not None:
+    dimred_file = args.dim_red
+if args.image is not None:
+    image_file = args.image
+if args.config is not None:
+    config_file = args.config
+
+n_clusters = args.n_clusters
+technology = args.technology
+seed = args.seed
+
+import anndata as ad
+import numpy as np
+import pandas as pd
+import scipy as sp
+from PIL import Image
+
+# ... or AnnData if you want
+def get_anndata(args):
+    import anndata as ad
+    import scipy as sp
+    from PIL import Image
+
+    X = sp.io.mmread(args.matrix)
+    if sp.sparse.issparse(X):
+        X = X.tocsr()
+
+    observations = pd.read_table(args.observations, index_col=0)
+    features = pd.read_table(args.features, index_col=0)
+
+    # Filter
+    if "selected" in observations.columns:
+        X = X[observations["selected"].to_numpy().nonzero()[0], :]
+        observations = observations.loc[lambda df: df["selected"]]
+    if "selected" in features.columns:
+        X = X[:, features["selected"].to_numpy().nonzero()[0]]
+        features = features.loc[lambda df: df["selected"]]
+
+    coordinates = (
+        pd.read_table(args.coordinates, index_col=0)
+        .loc[observations.index, :]
+        .to_numpy()
+    )
+
+    adata = ad.AnnData(
+        X=X, obs=observations, var=features, obsm={"spatial_pixel": coordinates}
+    )
+
+    if args.image is not None:
+        adata.uns["image"] = np.array(Image.open(args.image))
+    else:
+        adata.uns["image"] = None
+
+    return adata
+
+
+adata = get_anndata(args)
+adata.var_names_make_unique()
+
+
+# TODO set the seed, if the method requires the seed elsewhere please pass it on
+import random
+
+random.seed(seed)
+# np.random.seed(seed)
+# torch.manual_seed(seed)
+
+## Your code goes here
+import scanpy as sc
+import json
+
+# get the json config
+with open (args.config, "r") as c:
+    config = json.load(c)
+
+# scanpy settings
+# sc.settings.set_figure_params(scanpy=True, dpi=80, dpi_save=500, frameon=True, vector_friendly=True, fontsize=14)
+
+
+    # throw warning about not using the num_clusters as a parameter, because scanpy uses leiden or louvain and needs the resolution parameter, which is defined in the config.json. There exists a good function to perform an extensive search for the right resolution parameter to define the desired num_clusters, but we still have licensing issues. For more, see the SpaceHack2.0 GitHub issue #139
+import warnings
+
+warnings.warn("Scanpy uses leiden/louvain for clustering, which relies on the resolution parameter in the config file. The parameter num_clusters will be ignored.", UserWarning)
+
+# scanpy starts here
+sc.pp.normalize_total(adata)
+sc.pp.log1p(adata)
+sc.tl.pca(adata)
+sc.pp.neighbors(adata, n_neighbors = config["n_neighbors"], n_pcs=config["n_pcs"], random_state=seed)
```

Comment:
If you use the dim_red file as input you don't need to preprocess the counts nor run PCA.